File: src/test/java/com/runninglane/dto/buddy/test/java/cases/contract/ReturnedClassContractTest.java
package com.runninglane.dto.buddy.test.java.cases.contract;

import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import com.runninglane.dto.buddy.test.java.cases.contract.CompliantBaseClasses.*;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

/**
 * The rule is the returned class must be a concrete class
 */
class ReturnedClassContractTest {
    private final DtoBuddy dtoBuddy = new DtoBuddy();

    private void test(Class<?> baseClass, boolean expectBaseClassReturned) {
        try {
            Class<?> concreteClass = dtoBuddy.implement(baseClass);
            assertFalse(concreteClass.isInterface());
            assertFalse(java.lang.reflect.Modifier.isAbstract(concreteClass.getModifiers()));
            if (expectBaseClassReturned) {
                assertEquals(baseClass, concreteClass);
            }
        } catch (Throwable e) {
            throw new RuntimeException("Failed to test " + baseClass.getSimpleName(), e);
        }
    }

    @Test
    public void testInterface() {
        java.util.List.of(
            InterfaceWithNoMember.class,
            InterfaceWithAbstractGetter.class,
            InterfaceWithAbstractGetterAndAbstractSetter.class,
            InterfaceWithAbstractProperty.class
        ).forEach(baseClass -> test(baseClass, false));
    }

    @Test
    public void testAbstractClass() {
        java.util.List.of(
            AbstractClassWithNoMember.class,
            AbstractClassWithAbstractGetter.class,
            AbstractClassWithAbstractGetterAndAbstractSetter.class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithoutField.class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithField.class,
            AbstractClassWithAbstractImmutableProperty.class,
            AbstractClassWithAbstractMutableProperty.class,
            AbstractClassWithConcreteMutableProperty.class
        ).forEach(baseClass -> test(baseClass, false));
    }

    @Test
    public void testConcreteClass() {
        java.util.Map.of(
            ConcreteClassWithNoMember.class, true,
            ConcreteClassWithGetter.class, false,
            ConcreteClassWithSetter.class, false,
            ConcreteClassWithGetterAndSetterWithoutField.class, true,
            ConcreteClassWithGetterAndSetterWithField.class, true
        ).forEach((baseClass, expectBaseClassReturned) ->
            test(baseClass, expectBaseClassReturned));
    }

}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/contract/ReturnedClassContractTest.kt
package com.runninglane.dto.buddy.test.cases.contract

import com.runninglane.dto.buddy.DtoBuddy
import java.lang.reflect.Modifier
import kotlin.reflect.KClass
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse

/**
 * The rule is the returned class must be a concrete class
 */
class ReturnedClassContractTest {
    private val dtoBuddy = DtoBuddy()

    private fun test(baseClass: KClass<*>, expectBaseClassReturned: Boolean) {
        try {
            val concreteClass = dtoBuddy.implement(baseClass)
            assertFalse(concreteClass.java.isInterface)
            assertFalse(concreteClass.isAbstract)
            if (expectBaseClassReturned) {
                assertEquals(baseClass, concreteClass)
            }
        } catch (e: Throwable) {
            throw RuntimeException("Failed to test ${baseClass.simpleName}", e)
        }
    }

    @Test
    fun testInterface() {
        listOf(
            InterfaceWithNoMember::class,
            InterfaceWithAbstractGetter::class,
            InterfaceWithAbstractGetterAndAbstractSetter::class,
            InterfaceWithAbstractProperty::class
        ).forEach { baseClass ->
            test(baseClass, false)
        }
    }

    @Test
    fun testAbstractClass() {
        listOf(
            AbstractClassWithNoMember::class,
            AbstractClassWithAbstractGetter::class,
            AbstractClassWithAbstractGetterAndAbstractSetter::class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithoutField::class,
            AbstractClassWithConcreteGetterAndConcreteSetterWithField::class,
            AbstractClassWithAbstractImmutableProperty::class,
            AbstractClassWithAbstractMutableProperty::class,
            AbstractClassWithConcreteMutableProperty::class
        ).forEach { baseClass ->
            test(baseClass, false)
        }
    }

    @Test
    fun testConcreteClass() {
        listOf(
            ConcreteClassWithNoMember::class to true,
            ConcreteClassWithGetter::class to false,
            ConcreteClassWithSetter::class to false,
            ConcreteClassWithGetterAndSetterWithoutField::class to true,
            ConcreteClassWithGetterAndSetterWithField::class to true,
            ConcreteClassWithConcreteImmutableProperty::class to false,
            ConcreteClassWithConcreteMutableProperty::class to true
        ).forEach { (baseClass, expectBaseClassReturned) ->
            test(baseClass, expectBaseClassReturned)
        }
    }

}